state-chain/chains/src/sol/api.rs

```diff
@@ -1,70 +1,233 @@
-pub enum SolanaApi<Env> {
-	AllBatch {
-		fetch_params: vec::Vec<FetchAssetParams<Solana>>,
-		transfer_params: vec::Vec<TransferAssetParams<Solana>>,
-	},
-	SetAggKeyWithAggKey {
-		maybe_old_key: Option<<SolanaCrypto as ChainCrypto>::AggKey>,
-		new_key: <SolanaCrypto as ChainCrypto>::AggKey,
-	},
-	TransferFallback {
-		transfer_param: TransferAssetParams<Solana>,
-	},
-	ExecutexSwapAndCall {
+pub enum SolanaTransactionType {
+	BatchFetch,
+	Transfer,
+	RotateAggKey,
+}
+
+/// The Solana Api call. Contains a call_type and the actual Transaction itself.
+#[derive(CloneNoBound, DebugNoBound, PartialEqNoBound, EqNoBound, Encode, Decode, TypeInfo)]
+#[scale_info(skip_type_params(Environment))]
+pub struct SolanaApi<Environment: 'static> {
+	call_type: SolanaTransactionType,
+	transaction: SolTransaction,
+	#[doc(hidden)]
+	#[codec(skip)]
+	_phantom: PhantomData<Environment>,
+}
+
+impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
+	pub fn batch_fetch(
+		deposit_channels: Vec<DepositChannel<Solana>>,
+	) -> Result<Self, SolanaTransactionBuildingError> {
+		// Lookup the current Aggkey
+		let agg_key = Environment::lookup_account(SolanaEnvAccountLookupKey::AggKey)?;
+		let vault_program_data_account =
+			Environment::lookup_account(SolanaEnvAccountLookupKey::VaultProgramDataAccount)?;
+		let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
+			.expect("Preset System Program ID account must be valid.");
+		let next_nonce =
+			Environment::lookup_account(SolanaEnvAccountLookupKey::AvailableNonceAccount)?;
+		let compute_price = Environment::compute_price()?;
+
+		// Build the instruction_set
+		let instruction_set = SolanaInstructionBuilder::fetch_from(
+			deposit_channels,
+			vault_program_data_account,
+			system_program_id,
+			agg_key,
+			next_nonce,
+			compute_price,
+		);
+		let transaction =
+			SolTransaction::new_unsigned(SolMessage::new(&instruction_set, Some(&agg_key.into())));
+
+		Ok(Self {
+			call_type: SolanaTransactionType::BatchFetch,
+			transaction,
+			_phantom: Default::default(),
+		})
+	}
+
+	pub fn transfer(
 		transfer_param: TransferAssetParams<Solana>,
-		source_chain: cf_primitives::ForeignChain,
-		source_address: Option<ForeignChainAddress>,
-		gas_budget: <Solana as Chain>::ChainAmount,
-		message: vec::Vec<u8>,
-	},
-	ApiCall {
-		threshold_signature: <SolanaCrypto as ChainCrypto>::ThresholdSignature,
-	},
+	) -> Result<Self, SolanaTransactionBuildingError> {
+		// Lookup the current Aggkey
+		let agg_key = Environment::lookup_account(SolanaEnvAccountLookupKey::AggKey)?;
+		let next_nonce =
+			Environment::lookup_account(SolanaEnvAccountLookupKey::AvailableNonceAccount)?;
+		let compute_price = Environment::compute_price()?;
+
+		// Build the instruction_set
+		let instruction_set =
+			SolanaInstructionBuilder::transfer(transfer_param, agg_key, next_nonce, compute_price);
+
+		let transaction =
+			SolTransaction::new_unsigned(SolMessage::new(&instruction_set, Some(&agg_key.into())));
```

Comment:
we need to add the durable nonce here while creating the ```SolMessage``` using the function ```new_with_blockhash```. Moreover, the nonce_account and the hash inside it should be coupled in the type and for each nonce account, we should also store its nonce and then use it here.

Reply:
Updated this. Also made the SolMessage::new() test only. Other functions in the Message struct had its visibility updated as well. This way test-only function won't accidentally make its way into production/non-test.